gcp_variant_transforms/libs/conflicts_reporter.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2018 Google Inc.  All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generates conflicts report.
+
+The report is aimed to help the user to easily import the malformed/incompatible
+VCF files. It contains the conflicted header definitions, the corresponding file
+paths, and the suggested resolutions.
+TODO(yifangchen): Eventually, it also contains the inferred header definitions
+and the malformed records.
+"""
+
+from typing import Dict, List, Union  # pylint: disable=unused-import
+
+from apache_beam.io.filesystems import FileSystems
+
+from gcp_variant_transforms.beam_io.vcf_header_io import VcfHeader  # pylint: disable=unused-import
+from gcp_variant_transforms.beam_io.vcf_header_io import VcfParserHeaderKeyConstants
+from gcp_variant_transforms.transforms.merge_header_definitions import Definition  # pylint: disable=unused-import
+from gcp_variant_transforms.transforms.merge_header_definitions import VcfHeaderDefinitions  # pylint: disable=unused-import
+
+
+_HEADER_LINE = 'ID;Conflicts;Proposed Resolution\n'
+_NO_CONFLICTS_MESSAGE = 'No conflicts found.'
+_NO_SOLUTION_MESSAGE = 'Not resolved'
+
+
+def generate_conflicts_report(file_path,
+                              header_definitions,
+                              resolved_headers=None):
+  # type: (str, VcfHeaderDefinitions, VcfHeader) -> None
+  """Generates a report.
+
+  Combines the conflicts extracted from ``header_definitions`` and their
+  resolutions from ``resolved_headers`` to generate a conflicts report.
+  ``file_path`` specifies the location where the conflicts report is saved.
+  """
+  resolved_headers = resolved_headers or VcfHeader()
+  content_lines = []
+  content_lines.extend(_generate_contents(
+      _extract_conflicts(header_definitions.formats), resolved_headers.formats))
+  content_lines.extend(_generate_contents(
+      _extract_conflicts(header_definitions.infos), resolved_headers.infos))
+  _write_to_report(content_lines, file_path)
+
+
+def _extract_conflicts(
+    definitions  # type: Dict[str, Dict[Definition, List[str]]]
+):
```

Comment:
nit: should `):` be moved forward by 4space? can you check with style guide? Thanks!

Reply:
The style guide didn't say anything for this case. But I think leaving 4 spaces are more readable. It is also adopted in multiple modules in this project. Thanks for pointing out. Done.